narwhal/types/src/primary.rs

```diff
@@ -388,6 +446,7 @@ pub struct Certificate {
     aggregated_signature: AggregateSignature,
     #[serde_as(as = "NarwhalBitmap")]
     signed_authorities: roaring::RoaringBitmap,
+    pub metadata: Metadata,
```

Comment:
Adding `Metadata` / timestamp to all types may be controversial. If possible, it seems better to keep internal diagnostic data separated from application data. There are two alternatives I can think of:
1. Rely on anemo request metadata to pass along timestamps between validators. Use a separate function arg to pass timestamps within validators.
2. Log header digest / certificate digest / batch ID with timestamp, and calculate latencies offline.

Reply:
> Adding Metadata / timestamp to all types may be controversial. If possible, it seems better to keep internal diagnostic data separated from application data. There are two alternatives I can think of:
> 
> 1. Rely on anemo request metadata to pass along timestamps between validators. Use a separate function arg to pass timestamps within validators.
> 2. Log header digest / certificate digest / batch ID with timestamp, and calculate latencies offline.

Thanks @mwtian for the feedback on this. Regarding the diagnostic data kept separated from application data, I agree but find it a bit challenging to get this right or without triggering a wider refactoring. More specifically the way I treated `metadata` here are in the way that you would treat metadata in a common database record (that's what we do with them in the end of the day - store and process them) - so I see them close to the application level but as "metadata" data.

> 1. Rely on anemo request metadata to pass along timestamps between validators. Use a separate function arg to pass timestamps within validators.

I see the point on this. The challenge here is with nested components, ex Certificate includes the Header, so we somehow need to accommodate both timestamps. Also to make this happen we'll probably need to change all the points where we do handle Headers, Certificates and Batches and I wouldn't like to make that many changes just for that. But I would be interested to catch up more about this idea to make sure I am getting it correctly.

> 2. Log header digest / certificate digest / batch ID with timestamp, and calculate latencies offline.
To be honest I would avoid using logs to generate insights for operational data as logging make change at any point (ex sampling might be introduced etc). Also I am not quite sure how we would achieve reliable aggregated latency metrics (in Grafana) by just using logging (especially targeting matching records) - any pointers where should I look at?

Reply:
Maybe you could have the metadata field and its usage hidden behind a 'diagnostic' feature flag?

Reply:
Seeing as how this is the precursor to getting timestamps in NW ([Issue#5297](https://github.com/MystenLabs/sui/issues/5297)), we can leave the refactoring for a future PR